File: examples/standalone/ts/03-objects.ts
// Example 3: Object parameter with field access in conditions
// Tests shatter's ability to generate structured inputs and reason about field values.

interface UserProfile {
    name: string;
    age: number;
    isVerified: boolean;
    role: string;
}

// EXPECTED BRANCHES:
//   1. age < 0                        → throws Error("invalid age")
//   2. age < 13                       → "child"
//   3. age < 18                       → "teen"
//   4. age >= 18, isVerified, admin   → "admin"
//   5. age >= 18, isVerified, !admin  → "verified-user"
//   6. age >= 18, !isVerified         → "unverified-user"

export function categorizeUser(user: UserProfile): string {
    if (user.age < 0) {
        throw new Error("invalid age");
    }
    if (user.age < 13) {
        return "child";
    }
    if (user.age < 18) {
        return "teen";
    }
    if (user.isVerified) {
        if (user.role === "admin") {
            return "admin";
        }
        return "verified-user";
    }
    return "unverified-user";
}

interface Rectangle {
    width: number;
    height: number;
}

// EXPECTED BRANCHES:
//   1. width <= 0 OR height <= 0  → throws Error("non-positive dimension")
//   2. width === height            → "square"
//   3. area > 10000                → "large-rectangle"
//   4. area <= 10000               → "small-rectangle"

export function describeRectangle(rect: Rectangle): string {
    if (rect.width <= 0 || rect.height <= 0) {
        throw new Error("non-positive dimension");
    }
    if (rect.width === rect.height) {
        return "square";
    }
    const area = rect.width * rect.height;
    if (area > 10000) {
        return "large-rectangle";
    }
    return "small-rectangle";
}

File: examples/standalone/go/03-objects.go
package main

import "fmt"

// Example 3: Struct parameters with field access in conditions.
// Tests shatter's ability to generate structured inputs and reason about field values.

// UserProfile — test fixture for struct-field access analysis.
type UserProfile struct {
	Name       string
	Age        int
	IsVerified bool
	Role       string
}

// CategorizeUser — 6 branches: age<0→error, age<13→"child", age<18→"teen",
// age≥18+verified+admin→"admin", age≥18+verified→"verified-user", else→"unverified-user".
// Analyzer should detect nested field checks (age ranges, boolean, string equality).
func CategorizeUser(user UserProfile) (string, error) {
	if user.Age < 0 {
		return "", fmt.Errorf("invalid age")
	}
	if user.Age < 13 {
		return "child", nil
	}
	if user.Age < 18 {
		return "teen", nil
	}
	if user.IsVerified {
		if user.Role == "admin" {
			return "admin", nil
		}
		return "verified-user", nil
	}
	return "unverified-user", nil
}

// Rectangle — test fixture for dimension-based classification.
type Rectangle struct {
	Width  float64
	Height float64
}

// DescribeRectangle — 4 branches: non-positive dimension→error,
// width==height→"square", area>10000→"large-rectangle", else→"small-rectangle".
func DescribeRectangle(rect Rectangle) (string, error) {
	if rect.Width <= 0 || rect.Height <= 0 {
		return "", fmt.Errorf("non-positive dimension")
	}
	if rect.Width == rect.Height {
		return "square", nil
	}
	area := rect.Width * rect.Height
	if area > 10000 {
		return "large-rectangle", nil
	}
	return "small-rectangle", nil
}
